### PR TITLE
fix(gnovm): in op_binary, return typed booleans where appropriate

### PR DESCRIPTION
**Pull request by thehowl**
bool8.gno was failing, because the result of the `==` expression is an untyped boolean, while the first value is a typed boolean. This PR ensures that if either of the values in a binary expression is typed, we return a typed bool instead of an untyped bool.

### CODE REVIEW

**Comment by Gno2D2** [comment]
I'm a bot that assists the Gno Core team in maintaining this repository. My role is to ensure that contributors understand and follow our guidelines, helping to streamline the development process.

The following requirements must be fulfilled before a pull request can be merged.
Some requirement checks are automated and can be verified by the CI, while others need manual verification by a staff member.

These requirements are defined in this [configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

## Automated Checks

 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 The pull request head branch must be up-to-date with its base ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/keeping-your-pull-request-in-sync-with-the-base-branch))


## Manual Checks

*No manual checks match this pull request.*

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>The pull request head branch must be up-to-date with its base (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Head branch (thehowl:dev/morgan/runtime-typed-bools) is up to date with base (master): behind by 0 / ahead by 1

```
</blockquote></details>

</blockquote></details>



</blockquote></details>

**Comment by ltzmaxwell** [comment]
I think the root cause for this is in preprocess stage, the untyped T is not correctly converted into the corresponding typed T.

It happens in func `convertType`, when t is nil or `interfaceType`, should use `defaultTypeOf(t)` as the dest `t`.

File: gnovm/tests/files/bool8.gno
Comment (ltzmaxwell) [inline gnovm/tests/files/bool8.gno:9]: when calling isEqual, the second parameter(untyped bool) should be converted to `bool`, while the dest `t` is interface type.

File: gnovm/tests/files/bool8.gno
Comment (ltzmaxwell) [inline gnovm/tests/files/bool8.gno:12]: when calling `isEqual`, the second parameter (untyped bool) should be converted into `bool`, while the dest `t` is interface type.

**Comment by thehowl** [comment]
@ltzmaxwell thanks for the review!

I did some inspection and think I found a more correct fix. Can you take another look?

**Review by ltzmaxwell — APPROVED** [review]

**Comment by ltzmaxwell** [comment]
LGTM.

**Pull request merged**
